game-app/game-core/src/main/java/games/strategy/triplea/ui/SimpleUnitPanel.java

```diff
@@ -41,85 +41,31 @@ public enum Style {
-            } else if (n1 instanceof Resource) {
-              final Resource r1 = (Resource) n1;
-              if (n2 instanceof UnitType) {
-                // final UnitType u2 = (UnitType) n2;
-                return 1;
-              } else if (n2 instanceof Resource) {
-                final Resource r2 = (Resource) n2;
-                return r1.getName().compareTo(r2.getName());
-              } else {
-                return n1.getName().compareTo(n2.getName());
-              }
-            }
-            return n1.getName().compareTo(n2.getName());
+          var comparison = ruleComparator.compare(o1, o2);
+          if (comparison == 0) {
+            return o1.getName().compareTo(o2.getName());
+          } else {
+            return comparison;
           }
-          if (o1.getResults().size() > o2.getResults().size()) {
-            return -1;
-          } else if (o1.getResults().size() < o2.getResults().size()) {
-            return 1;
-          }
-          return o1.getName().compareTo(o2.getName());
         }
       };
 
   private final Comparator<RepairRule> repairRuleComparator =
       new Comparator<>() {
-        final UnitTypeComparator utc = new UnitTypeComparator();
+        final RuleComparator ruleComparator = new RuleComparator();
 
         @Override
         public int compare(final RepairRule o1, final RepairRule o2) {
-          if (o1.getResults().size() == 1 && o2.getResults().size() == 1) {
-            final NamedAttachable n1 = o1.getResults().keySet().iterator().next();
-            final NamedAttachable n2 = o2.getResults().keySet().iterator().next();
-            if (n1 instanceof UnitType) {
-              final UnitType u1 = (UnitType) n1;
-              if (n2 instanceof UnitType) {
-                final UnitType u2 = (UnitType) n2;
-                return utc.compare(u1, u2);
-              } else if (n2 instanceof Resource) {
-                // final Resource r2 = (Resource) n2;
-                return -1;
-              }
-              return n1.getName().compareTo(n2.getName());
-            } else if (n1 instanceof Resource) {
-              final Resource r1 = (Resource) n1;
-              if (n2 instanceof UnitType) {
-                // final UnitType u2 = (UnitType) n2;
-                return 1;
-              } else if (n2 instanceof Resource) {
-                final Resource r2 = (Resource) n2;
-                return r1.getName().compareTo(r2.getName());
-              } else {
-                return n1.getName().compareTo(n2.getName());
-              }
-            }
-            return n1.getName().compareTo(n2.getName());
-          }
-          if (o1.getResults().size() > o2.getResults().size()) {
-            return -1;
-          } else if (o1.getResults().size() < o2.getResults().size()) {
-            return 1;
+          var comparison = ruleComparator.compare(o1, o2);
+          if (comparison == 0) {
+            return o1.getName().compareTo(o2.getName());
```

Comment:
If `RuleComparator` was made generic, i.e. the class definition changed to
```java
public class RuleComparator<T extends Rule> implements Comparator<T> {
  @Override
  public int compare(T o1, T o2) {
    // your code here
  }
}
```

there would be a super simple way to shorten this code like this:
```java
private final Comparator<RepairRule> repairRuleComparator = new RuleComparator<RepairRule>().thenComparing(RepairRule::getName);
```

Alternatively you might be able to extend the Rule interface to include `getName()`, not sure if that collides with the `DefaultNamed` class though. But if it works you could merge `productionRuleComparator` and `repairRuleComparator` into a single comparator without any generics, just a simple interface declaration

Reply:
Also if you're able to unify those two classes into a single interface you could probably just move the final comparison into the actual class I think